catalyst/dl/core/state.py

```diff
@@ -87,8 +90,9 @@ def __init__(
         self.timer = TimerManager()
 
         # base metrics
-        self.lr = None
-        self.momentum = None
+        single_optimizer = isinstance(optimizer, Optimizer)
+        self.lr = None if single_optimizer else defaultdict(lambda: None)
```

Comment:
dont use `defaultdict`, try https://github.com/catalyst-team/safitty from catalyst-ecosystem

Reply:
So far I believe defaultdicts are much more clear here with much less code.
I agree that dicts instead may prevent some errors (but without safitty, again)

Maybe you can give me a masterclass?
```python
        self.lr = None
        self.momentum = None
        if isinstance(optimizer, dict):
            self.lr = dict()
            self.momentum = dict()
            for key in optimizer:
                safitty.set(self.lr, key, value=None)
                safitty.set(self.momentum, key, value=None)
```